File: CloneGraph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        def dfs(parent):
            if parent==None:
                return None
            
            visited[parent]=Node(parent.val)
            for child in parent.neighbors:
                if child not in visited:                
                    visited[parent].neighbors.append(dfs(child))
                else:
                    visited[parent].neighbors.append(visited[child])
                
            return visited[parent]
        
        visited={}
        return dfs(node)
    
    

File: CloneGraph.cpp
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* cloneGraph(Node* node) {
        if(!node)
            return NULL;
        if(m.find(node)==m.end()){
            m[node]=new Node(node->val);
            for(int i=0;i<node->neighbors.size();i++)
                m[node]->neighbors.push_back(cloneGraph(node->neighbors[i]));
        }
        
        return m[node];
    }
    
private:
    unordered_map<Node*,Node*>m;
};
